File: src/backend/report_options_backend.py
#!/usr/bin/env python3
"""
Report options backend: report layout settings and order selection.
"""

from qt_compat import QObject, pyqtProperty, pyqtSignal, pyqtSlot

from utils.qsettings_store import QSettingsStore


_SETTINGS_KEY_INCLUDE_CLIENT_NAME = "report_options/include_client_name"
_SETTINGS_KEY_INCLUDE_CLIENT_ADDRESS = "report_options/include_client_address"
_SETTINGS_KEY_INCLUDE_REPORT_DATE = "report_options/include_report_date"
_SETTINGS_KEY_INCLUDE_PERSONAL_INFO = "report_options/include_personal_info"
_SETTINGS_KEY_INCLUDE_REPORT_HEADER = "report_options/include_report_header"
_SETTINGS_KEY_REPORT_HEADER_TEXT = "report_options/report_header_text"
_SETTINGS_KEY_GROUP_BY_SUBOBJECTS = "report_options/group_by_subobjects"

_DEFAULT_REPORT_HEADER = "Отчет о проделанной работе по установке сантехники"


class ReportOptionsBackend(QObject):
    """Backend for report generation options and selected orders."""

    optionsChanged = pyqtSignal()
    orderSelectionChanged = pyqtSignal()

    def __init__(self, parent=None):
        super().__init__(parent)
        self._include_client_name = True
        self._include_client_address = True
        self._include_report_date = True
        self._include_personal_info = True
        self._include_report_header = True
        self._report_header_text = _DEFAULT_REPORT_HEADER
        self._group_by_subobjects = True
        self._selected_orders = set()
        self._load_settings()

    def _load_settings(self):
        with QSettingsStore() as store:
            self._include_client_name = store.read(
                _SETTINGS_KEY_INCLUDE_CLIENT_NAME, True, value_type=bool
            )
            self._include_client_address = store.read(
                _SETTINGS_KEY_INCLUDE_CLIENT_ADDRESS, True, value_type=bool
            )
            self._include_report_date = store.read(
                _SETTINGS_KEY_INCLUDE_REPORT_DATE, True, value_type=bool
            )
            self._include_personal_info = store.read(
                _SETTINGS_KEY_INCLUDE_PERSONAL_INFO, True, value_type=bool
            )
            self._include_report_header = store.read(
                _SETTINGS_KEY_INCLUDE_REPORT_HEADER, True, value_type=bool
            )
            self._report_header_text = store.read(
                _SETTINGS_KEY_REPORT_HEADER_TEXT,
                _DEFAULT_REPORT_HEADER,
                value_type=str,
            ) or _DEFAULT_REPORT_HEADER
            self._group_by_subobjects = store.read(
                _SETTINGS_KEY_GROUP_BY_SUBOBJECTS, True, value_type=bool
            )

    def _persist_settings(self):
        with QSettingsStore() as store:
            store.write(_SETTINGS_KEY_INCLUDE_CLIENT_NAME, self._include_client_name)
            store.write(_SETTINGS_KEY_INCLUDE_CLIENT_ADDRESS, self._include_client_address)
            store.write(_SETTINGS_KEY_INCLUDE_REPORT_DATE, self._include_report_date)
            store.write(_SETTINGS_KEY_INCLUDE_PERSONAL_INFO, self._include_personal_info)
            store.write(_SETTINGS_KEY_INCLUDE_REPORT_HEADER, self._include_report_header)
            store.write(_SETTINGS_KEY_REPORT_HEADER_TEXT, self._report_header_text)
            store.write(_SETTINGS_KEY_GROUP_BY_SUBOBJECTS, self._group_by_subobjects)

    def as_dict(self):
        """Return current options as plain dict for report builder."""
        return {
            "include_client_name": self._include_client_name,
            "include_client_address": self._include_client_address,
            "include_report_date": self._include_report_date,
            "include_personal_info": self._include_personal_info,
            "include_report_header": self._include_report_header,
            "report_header_text": self._report_header_text,
            "group_by_subobjects": self._group_by_subobjects,
        }

    @pyqtProperty(bool, notify=optionsChanged)
    def includeClientName(self):
        return self._include_client_name

    @includeClientName.setter
    def includeClientName(self, value):
        if self._include_client_name != value:
            self._include_client_name = value
            self.optionsChanged.emit()

    @pyqtProperty(bool, notify=optionsChanged)
    def includeClientAddress(self):
        return self._include_client_address

    @includeClientAddress.setter
    def includeClientAddress(self, value):
        if self._include_client_address != value:
            self._include_client_address = value
            self.optionsChanged.emit()

    @pyqtProperty(bool, notify=optionsChanged)
    def includeReportDate(self):
        return self._include_report_date

    @includeReportDate.setter
    def includeReportDate(self, value):
        if self._include_report_date != value:
            self._include_report_date = value
            self.optionsChanged.emit()

    @pyqtProperty(bool, notify=optionsChanged)
    def includePersonalInfo(self):
        return self._include_personal_info

    @includePersonalInfo.setter
    def includePersonalInfo(self, value):
        if self._include_personal_info != value:
            self._include_personal_info = value
            self.optionsChanged.emit()

    @pyqtProperty(bool, notify=optionsChanged)
    def includeReportHeader(self):
        return self._include_report_header

    @includeReportHeader.setter
    def includeReportHeader(self, value):
        if self._include_report_header != value:
            self._include_report_header = value
            self.optionsChanged.emit()

    @pyqtProperty(str, notify=optionsChanged)
    def reportHeaderText(self):
        return self._report_header_text

    @reportHeaderText.setter
    def reportHeaderText(self, value):
        text = value or ""
        if self._report_header_text != text:
            self._report_header_text = text
            self.optionsChanged.emit()

    @pyqtProperty(bool, notify=optionsChanged)
    def groupBySubobjects(self):
        return self._group_by_subobjects

    @groupBySubobjects.setter
    def groupBySubobjects(self, value):
        if self._group_by_subobjects != value:
            self._group_by_subobjects = value
            self.optionsChanged.emit()

    @pyqtSlot(result=bool)
    def saveSettings(self):
        """Persist static report options to application settings."""
        self._persist_settings()
        return True

    @pyqtSlot()
    def clearOrderSelection(self):
        """Drop all selected orders (before report options dialog opens)."""
        if not self._selected_orders:
            return
        self._selected_orders.clear()
        self.orderSelectionChanged.emit()

    @pyqtSlot(int, result=bool)
    def orderSelected(self, start_order_at):
        return int(start_order_at) in self._selected_orders

    @pyqtSlot(int, bool)
    def setOrderSelected(self, start_order_at, selected):
        value = int(start_order_at)
        if selected:
            if value not in self._selected_orders:
                self._selected_orders.add(value)
                self.orderSelectionChanged.emit()
            return

        if value in self._selected_orders:
            self._selected_orders.remove(value)
            self.orderSelectionChanged.emit()

    @pyqtSlot(result=list)
    def selectedOrderTimestamps(self):
        return self._selected_orders

File: cpp/backend/ReportOptionsBackend.cpp
#include "ReportOptionsBackend.h"

#include "QSettingsStore.h"

namespace workorder
{

namespace
{

const char *kIncludeClientName = "report_options/include_client_name";
const char *kIncludeClientAddress = "report_options/include_client_address";
const char *kIncludeReportDate = "report_options/include_report_date";
const char *kIncludePersonalInfo = "report_options/include_personal_info";
const char *kIncludeReportHeader = "report_options/include_report_header";
const char *kReportHeaderText = "report_options/report_header_text";
const char *kGroupBySubobjects = "report_options/group_by_subobjects";
const char *kDefaultReportHeader = "Отчет о проделанной работе по установке сантехники";

} // namespace

ReportOptionsBackend::ReportOptionsBackend(QObject *parent)
    : QObject(parent)
    , m_reportHeaderText(QString::fromUtf8(kDefaultReportHeader))
{
    loadSettings();
}

ReportOptionsBackend::~ReportOptionsBackend() = default;

bool ReportOptionsBackend::includeClientName() const { return m_includeClientName; }
void ReportOptionsBackend::setIncludeClientName(bool value)
{
    if(m_includeClientName == value)
        return;
    m_includeClientName = value;
    emit optionsChanged();
}

bool ReportOptionsBackend::includeClientAddress() const { return m_includeClientAddress; }
void ReportOptionsBackend::setIncludeClientAddress(bool value)
{
    if(m_includeClientAddress == value)
        return;
    m_includeClientAddress = value;
    emit optionsChanged();
}

bool ReportOptionsBackend::includeReportDate() const { return m_includeReportDate; }
void ReportOptionsBackend::setIncludeReportDate(bool value)
{
    if(m_includeReportDate == value)
        return;
    m_includeReportDate = value;
    emit optionsChanged();
}

bool ReportOptionsBackend::includePersonalInfo() const { return m_includePersonalInfo; }
void ReportOptionsBackend::setIncludePersonalInfo(bool value)
{
    if(m_includePersonalInfo == value)
        return;
    m_includePersonalInfo = value;
    emit optionsChanged();
}

bool ReportOptionsBackend::includeReportHeader() const { return m_includeReportHeader; }
void ReportOptionsBackend::setIncludeReportHeader(bool value)
{
    if(m_includeReportHeader == value)
        return;
    m_includeReportHeader = value;
    emit optionsChanged();
}

QString ReportOptionsBackend::reportHeaderText() const { return m_reportHeaderText; }
void ReportOptionsBackend::setReportHeaderText(const QString &value)
{
    if(m_reportHeaderText == value)
        return;
    m_reportHeaderText = value;
    emit optionsChanged();
}

bool ReportOptionsBackend::groupBySubobjects() const { return m_groupBySubobjects; }
void ReportOptionsBackend::setGroupBySubobjects(bool value)
{
    if(m_groupBySubobjects == value)
        return;
    m_groupBySubobjects = value;
    emit optionsChanged();
}

QVariantMap ReportOptionsBackend::asDict() const
{
    QVariantMap map;
    map.insert("include_client_name", m_includeClientName);
    map.insert("include_client_address", m_includeClientAddress);
    map.insert("include_report_date", m_includeReportDate);
    map.insert("include_personal_info", m_includePersonalInfo);
    map.insert("include_report_header", m_includeReportHeader);
    map.insert("report_header_text", m_reportHeaderText);
    map.insert("group_by_subobjects", m_groupBySubobjects);
    return map;
}

bool ReportOptionsBackend::saveSettings()
{
    persistSettings();
    return true;
}

void ReportOptionsBackend::clearOrderSelection()
{
    if(m_selectedOrders.isEmpty())
        return;
    m_selectedOrders.clear();
    emit orderSelectionChanged();
}

bool ReportOptionsBackend::orderSelected(int startOrderAt) const
{
    return m_selectedOrders.contains(startOrderAt);
}

void ReportOptionsBackend::setOrderSelected(int startOrderAt, bool selected)
{
    if(selected)
    {
        if(!m_selectedOrders.contains(startOrderAt))
        {
            m_selectedOrders.insert(startOrderAt);
            emit orderSelectionChanged();
        }
        return;
    }

    if(m_selectedOrders.remove(startOrderAt))
        emit orderSelectionChanged();
}

QVariantList ReportOptionsBackend::selectedOrderTimestamps() const
{
    QVariantList result;
    for(int value : m_selectedOrders)
        result.append(value);
    return result;
}

void ReportOptionsBackend::loadSettings()
{
    QSettingsStore store;
    m_includeClientName = store.read(kIncludeClientName, true).toBool();
    m_includeClientAddress = store.read(kIncludeClientAddress, true).toBool();
    m_includeReportDate = store.read(kIncludeReportDate, true).toBool();
    m_includePersonalInfo = store.read(kIncludePersonalInfo, true).toBool();
    m_includeReportHeader = store.read(kIncludeReportHeader, true).toBool();
    m_reportHeaderText = store.read(kReportHeaderText, m_reportHeaderText).toString();
    if(m_reportHeaderText.isEmpty())
        m_reportHeaderText = QString::fromUtf8(kDefaultReportHeader);
    m_groupBySubobjects = store.read(kGroupBySubobjects, true).toBool();
}

void ReportOptionsBackend::persistSettings()
{
    QSettingsStore store;
    store.write(kIncludeClientName, m_includeClientName);
    store.write(kIncludeClientAddress, m_includeClientAddress);
    store.write(kIncludeReportDate, m_includeReportDate);
    store.write(kIncludePersonalInfo, m_includePersonalInfo);
    store.write(kIncludeReportHeader, m_includeReportHeader);
    store.write(kReportHeaderText, m_reportHeaderText);
    store.write(kGroupBySubobjects, m_groupBySubobjects);
    store.sync();
}

} // namespace workorder
